File: backend/routes/songs.py
from flask import Blueprint, jsonify
from models import get_all_songs, get_song_by_id, scan_songs
from storage import serve_audio, get_signed_url

songs_bp = Blueprint('songs', __name__)


@songs_bp.route('', methods=['GET'])
def list_songs():
    songs = get_all_songs()
    return jsonify(songs)


@songs_bp.route('/refresh', methods=['POST'])
def refresh_songs():
    scan_songs()
    songs = get_all_songs()
    return jsonify({'message': f'Scanned {len(songs)} songs', 'count': len(songs)})


@songs_bp.route('/<int:song_id>', methods=['GET'])
def get_song(song_id):
    song = get_song_by_id(song_id)
    if not song:
        return jsonify({'error': 'Song not found'}), 404
    return jsonify(song)


@songs_bp.route('/<int:song_id>/stream', methods=['GET'])
def stream_song(song_id):
    song = get_song_by_id(song_id)
    if not song:
        return jsonify({'error': 'Song not found'}), 404
    return serve_audio(song['filename'])


@songs_bp.route('/<int:song_id>/stream-url', methods=['GET'])
def stream_url(song_id):
    song = get_song_by_id(song_id)
    if not song:
        return jsonify({'error': 'Song not found'}), 404
    url = get_signed_url(song['filename'])
    if url:
        return jsonify({'url': url})
    return jsonify({'url': f'/api/songs/{song_id}/stream'})

File: frontend/src/api/songs.js
import api from './index.js'

export async function getSongs() {
  const { data } = await api.get('/songs')
  return data
}

export async function getSong(id) {
  const { data } = await api.get(`/songs/${id}`)
  return data
}

export function getStreamUrl(id) {
  return `/api/songs/${id}/stream`
}

export async function getStreamSignedUrl(id) {
  const { data } = await api.get(`/songs/${id}/stream-url`)
  return data.url
}

export async function refreshSongs() {
  const { data } = await api.post('/songs/refresh')
  return data
}
